### PR TITLE
Ship hwlatdetect and cyclictest as part of cnf-tests image

### PR DESCRIPTION
**Pull request by Tal-or**
This update will allow the user/admin to run all latency tests using the same container image - cnf-tests image

### CODE REVIEW

**Comment by Tal-or** [comment]
/hold 
Depend on PR: https://github.com/openshift-kni/performance-addon-operators/pull/664

File: cnf-tests/README.md
Comment (yuvalk) [inline cnf-tests/README.md:114]: please add these new env vars as well as info on how to run to
https://github.com/openshift-kni/cnf-features-deploy/blob/master/cnf-tests/README.md#running-latency-test

File: cnf-tests/Dockerfile.openshift
Comment (yuvalk) [inline cnf-tests/Dockerfile.openshift:62]: why do you need another builder, which is identical to `builder` above?

Comment (Tal-or) [reply cnf-tests/Dockerfile.openshift]: Done

**Comment by fromanirh** [comment]
@Tal-or please check the latest commit here, we don't usually push merge commits

**Comment by Tal-or** [comment]
> @Tal-or please check the latest commit here, we don't usually push merge commits

Fixed! sorry for that

**Comment by fromanirh** [comment]
```
 STEP 1: FROM openshift/origin-release:golang-1.13 AS builder
time="2021-06-29T07:46:16Z" level=warning msg="failed, retrying in 5s ... (1/2). Error: Error initializing source docker://openshift/origin-release:golang-1.13: Error reading manifest golang-1.13 in docker.io/openshift/origin-release: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit"
time="2021-06-29T07:46:51Z" level=warning msg="failed, retrying in 5s ... (2/2). Error: Error initializing source docker://openshift/origin-release:golang-1.13: Error reading manifest golang-1.13 in docker.io/openshift/origin-release: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit"
STEP 2: FROM quay.io/openshift/origin-oc-rpms:4.8 AS oc 
```
another day, another quota

**Comment by fromanirh** [comment]
slow infra?
```
ClusterVersion: Installing "4.9.0-0.ci.test-2021-06-30-132215-ci-op-x9f8j6ql-latest" for 40 minutes: Working towards 4.9.0-0.ci.test-2021-06-30-132215-ci-op-x9f8j6ql-latest: 561 of 696 done (80% complete)
```

**Comment by Tal-or** [comment]
/retest
Looks like infra issues

**Comment by yuvalk** [comment]
investigating these failures, yield a gatekeeper problem:
```
2021-07-04T07:28:40.640397188Z {"level":"error","ts":1625383720.6392138,"logger":"controllers.operatorcondition","msg":"Error ensuring OperatorCondition Deployment EnvVars","request":"openshift-operators/gatekeeper-operator-product.v0.1.2-0.1624366317.p","error":"Deployment.apps \"gatekeeper-operator-controller-manager\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/build/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:293\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/build/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:248\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1\n\t/build/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1\n\t/build/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/build/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/build/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/build/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext\n\t/build/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.UntilWithContext\n\t/build/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:99"}
```

and
```
2021-07-04T07:28:44.663097733Z time="2021-07-04T07:28:44Z" level=info msg="couldn't ensure RBAC in target namespaces" csv=gatekeeper-operator-product.v0.1.2 error="no owned roles found" id=ya4Ni namespace=openshift-operators phase=Pending
```

both from OLM logs

IDK why this happens and more imporantly why it's flaky
@fedepaol

**Comment by yuvalk** [comment]
opened https://bugzilla.redhat.com/show_bug.cgi?id=1979019 to track the issue

**Comment by Tal-or** [comment]
/retest
After PR#590 merged

**Comment by Tal-or** [comment]
/retest
Seems like infra issue again

**Comment by fromanirh** [comment]
```
error: build error: failed to pull image: After retrying 2 times, Pull image still failed due to error: Error initializing source docker://golang:1.13: Error reading manifest 1.13 in docker.io/library/golang: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
```
:\

File: cnf-tests/Dockerfile
Comment (yanirq) [inline cnf-tests/Dockerfile:34]: Can you switch this to golang 1.16 (or above) ?

Comment (yanirq) [reply cnf-tests/Dockerfile:34]: maybe we should do this for all in a different PR

File: cnf-tests/Dockerfile
Comment (yanirq) [inline cnf-tests/Dockerfile:66]: I take it that hw latency can only run with python3 ?

File: cnf-tests/docgen/e2e.json
Comment (yanirq) [inline cnf-tests/docgen/e2e.json:38]: where is this coming from ?

File: cnf-tests/Dockerfile
Comment (fromanirh) [reply cnf-tests/Dockerfile:34]: +1 for bump to golang 1.16 and doing it in a separate PR

File: cnf-tests/Dockerfile
Comment (fromanirh) [reply cnf-tests/Dockerfile:66]: +1 can we clarify why this is needed now?

Comment (Tal-or) [reply cnf-tests/Dockerfile:66]: I didn't check with python2, but yes we need python for that

File: cnf-tests/docgen/e2e.json
Comment (Tal-or) [reply cnf-tests/docgen/e2e.json:38]: Automatically generated after I ran ```make check-tests-nodesc``` and ```make generate-cnf-tests-doc```

File: cnf-tests/Dockerfile
Comment (Tal-or) [reply cnf-tests/Dockerfile:34]: So you want me to change it here or on a separate PR?

Comment (fromanirh) [reply cnf-tests/Dockerfile:34]: separate PR please

File: cnf-tests/Dockerfile
Comment (fromanirh) [reply cnf-tests/Dockerfile:66]: ok, and which package requires python? just a comment here is enough.

**Review by fromanirh — APPROVED** [review]
/lgtm
/hold
waiting for other reviewers

**Comment by cynepco3hahue** [comment]
/lgtm

@Tal-or Did you test it locally?

**Comment by Tal-or** [comment]
> /lgtm
> 
> @Tal-or Did you test it locall

> /lgtm
> 
> @Tal-or Did you test it locally?

Yes. The runner's binaries are working

**Comment by fromanirh** [comment]
/hold cancel
IIRC we need to invoke these tests explicitely, so it's unlikely we will break `cnf-tests`.

**Comment by Tal-or** [comment]
> /hold cancel
> IIRC we need to invoke these tests explicitely, so it's unlikely we will break `cnf-tests`.

Why you're being such a pessimist? :)

**Comment by yanirq** [comment]
please check e2e failures here. the are several and ideally they all should pass (unless im missing something here)

**Comment by Tal-or** [comment]
> please check e2e failures here. there are several and ideally they all should pass (unless im missing something here)

I don't think so, I saw these types of errors in the past. I don't know the main reason behind it, but ```/retest``` usually solves that.
@fromanirh @yuvalk , do you mind having a look?

**Comment by Tal-or** [comment]
/retest
The test is failing because of the issue of empty strings which is explored independently.

**Comment by Tal-or** [comment]
/hold
Found a bug under the consumed tests

**Comment by Tal-or** [comment]
/hold cancel
Bug was fixed and merged to PAO's test https://github.com/openshift-kni/performance-addon-operators/pull/687

**Comment by Tal-or** [comment]
/retest
``` Latest error: Error setting IAM policy for project "**********************": googleapi: Error 409: There were concurrent policy changes``` infra issue?

**Review by fromanirh — COMMENTED** [review]
/lgtm
/hold
it would be nice to have cnf-features-deploy maintainers approval

**Review by MarSik — APPROVED** [review]
Generally this looks good to me, but someone else must approve.

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (fedepaol) [inline cnf-tests/pod-utils/cyclictest-runner/main.go:18]: No need to use var, you can use the short declaration form here rtPriority := ..

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (fedepaol) [inline cnf-tests/pod-utils/cyclictest-runner/main.go:19]: you can use flag.Duration here

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (fedepaol) [inline cnf-tests/pod-utils/cyclictest-runner/main.go:20]: same, if this is a number use flag.Int

File: cnf-tests/pod-utils/utils.go
Comment (fedepaol) [inline cnf-tests/pod-utils/utils.go:1]: utils packages are an antipattern in Go. Also, underscore in the package name is not idiomatic.

I'd suggest to use a package with a name like `node` and rename `PrintNodeInformation` to just `PrintInformation`

more info in
https://dave.cheney.net/2019/01/08/avoid-package-names-like-base-util-or-common
https://blog.golang.org/package-names

File: cnf-tests/pod-utils/utils.go
Comment (fedepaol) [inline cnf-tests/pod-utils/utils.go:22]: Can you extend this error so it's clear where it's coming from?

File: cnf-tests/pod-utils/utils.go
Comment (fedepaol) [inline cnf-tests/pod-utils/utils.go:31]: Is the difference in behaviour (fatal vs error) between this and GetSelfCPUs done purposely?
I believe Fatalf panics too.
I'd keep both consistent returning an error and exiting from main

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (fedepaol) [inline cnf-tests/pod-utils/cyclictest-runner/main.go:43]: This will cause a segfault if one of these parameters is not set.
If we need all it's common practice to check against their presence and print out flag.Usage()

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (fedepaol) [inline cnf-tests/pod-utils/cyclictest-runner/main.go:61]: Nit: extra line

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (fedepaol) [inline cnf-tests/pod-utils/cyclictest-runner/main.go:23]: this can be a duration too

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (fedepaol) [inline cnf-tests/pod-utils/cyclictest-runner/main.go:22]: this can be an int

File: cnf-tests/pod-utils/utils.go
Comment (fedepaol) [reply cnf-tests/pod-utils/utils.go:1]: Also, I'd suggest to create it under a pkg subdir so it's clear the others are executables and this is a package

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (Tal-or) [reply cnf-tests/pod-utils/cyclictest-runner/main.go:20]: This flag can also be an empty string ```""```

Comment (fedepaol) [reply cnf-tests/pod-utils/cyclictest-runner/main.go:20]: Can't you use 0 as default?

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (Tal-or) [reply cnf-tests/pod-utils/cyclictest-runner/main.go:19]: Basically, all the flags should eventually be converted into strings, but I agree that this emphasis the purpose of each variable a little bit.

Comment (fedepaol) [reply cnf-tests/pod-utils/cyclictest-runner/main.go:19]: Ah, did not notice. Then I guess it's fine and it's fine to rely on the cyclictestBinary to verify their correctness

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (fedepaol) [reply cnf-tests/pod-utils/cyclictest-runner/main.go:43]: Right, ignore what I just wrote

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (Tal-or) [reply cnf-tests/pod-utils/cyclictest-runner/main.go]: Empty means all threads, as you can see in the description

File: cnf-tests/pod-utils/node/node.go
Comment (fedepaol) [inline cnf-tests/pod-utils/node/node.go:1]: Can you move node under pod-utils/pkg?

File: cnf-tests/pod-utils/hwlatdetect-runner/main.go
Comment (fedepaol) [inline cnf-tests/pod-utils/hwlatdetect-runner/main.go:34]: If this is already a duration, can't you just use the value?

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (fedepaol) [inline cnf-tests/pod-utils/cyclictest-runner/main.go:42]: Can you just use the value as duration?

**Comment by fedepaol** [comment]
Can you squash the last changes in the relevant commit ?

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (fedepaol) [reply cnf-tests/pod-utils/cyclictest-runner/main.go]: I can see the description, but it feels totally odd.
If you want a default, you can set it to -1 and still achieve the result you want, or even better fetch maxCpus before setting the flags and use that as default value, I think that might work

Comment (yuvalk) [reply cnf-tests/pod-utils/cyclictest-runner/main.go]: > I can see the description, but it feels totally odd.
> If you want a default, you can set it to -1 and still achieve the result you want, or even better fetch maxCpus before setting the flags and use that as default value, I think that might work

IMHO

default should be 1 thread

to designate running on all cpus,
can either use 0 or the word MAX

the empty notion feels too weird

File: cnf-tests/pod-utils/node/node.go
Comment (Tal-or) [reply cnf-tests/pod-utils/node/node.go:1]: Just to make sure, the hierarchy should look like:
```
pod-utils
|_
    pkg
    |_
       node.go
```
and what ```node.go``` should be part of what ```package```?

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (Tal-or) [reply cnf-tests/pod-utils/cyclictest-runner/main.go]: I just mimic the behavior of the tool itself (also the description) but yes, I can totally make it more intuitive

Comment (yuvalk) [reply cnf-tests/pod-utils/cyclictest-runner/main.go]: > I just mimic the behavior of the tool itself (also the description) but yes, I can totally make it more intuitive

That's a good conflict though.
whether we want our tool to be more intuitive or more closely mimic the underlying tool ?

How close are the rest of the parameters?

Comment (Tal-or) [reply cnf-tests/pod-utils/cyclictest-runner/main.go]: The parameters descriptions are almost identical and also describe the default values the tool has.
There were other parameters that I added to the runner as "best-practices" that I saw being used in other projects, for example ```--histogram```

Comment (yuvalk) [reply cnf-tests/pod-utils/cyclictest-runner/main.go]: > The parameters descriptions are almost identical and also describe the default values the tool has.
> There were other parameters that I added to the runner as "best-practices" that I saw being used in other projects, for example `--histogram`

so in that case, I think I vote for preserving the original flags and meaning. our description can be less cryptic if needed.
and maybe work upstream on changing them ?

Comment (yuvalk) [reply cnf-tests/pod-utils/cyclictest-runner/main.go]: yes, if we believe `-t 0` is more intuitive and useful than `-t` we should talk with tool maintainers and see what they think if/when they fix it, we can do the same

Comment (Tal-or) [reply cnf-tests/pod-utils/cyclictest-runner/main.go]: Just to clarify, for now, I change nothing in order to be aligned with the tool behavior

File: cnf-tests/pod-utils/node/node.go
Comment (fedepaol) [reply cnf-tests/pod-utils/node/node.go]: Yes exactly. That's because all the other dirs under pod-utils are binaries

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (fedepaol) [reply cnf-tests/pod-utils/cyclictest-runner/main.go]: We are mixing two things. The type of threads, and the behaviour we want.
If it's a number, the type should be int. 

If we want to have a specific behaviour for when the parameter is not passed, then we can have a default as 0 or -1 and react according to that (the same way we do with `""`)

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (fedepaol) [inline cnf-tests/pod-utils/cyclictest-runner/main.go:33]: Isn't this different from what the comment says? If we don't pass `-t` then it will be maxCpus I guess?

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (fedepaol) [inline cnf-tests/pod-utils/cyclictest-runner/main.go:41]: the flag is now a duration, no need to multiply by second

File: cnf-tests/pod-utils/hwlatdetect-runner/main.go
Comment (fedepaol) [inline cnf-tests/pod-utils/hwlatdetect-runner/main.go:40]: This is a duration, if you want to fetch the seconds you need to use window.Seconds() otherwise it will be rendered like `52s`

File: cnf-tests/pod-utils/hwlatdetect-runner/main.go
Comment (fedepaol) [inline cnf-tests/pod-utils/hwlatdetect-runner/main.go:41]: same

File: cnf-tests/pod-utils/hwlatdetect-runner/main.go
Comment (fedepaol) [inline cnf-tests/pod-utils/hwlatdetect-runner/main.go:54]: nit: extra empty lines

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (Tal-or) [reply cnf-tests/pod-utils/cyclictest-runner/main.go:33]: this is how the help from the ```cyclictest``` looks like:
```
-t       --threads         one thread per available processor
-t [NUM] --threads=NUM     number of threads:
                           without NUM, threads = max_cpus
                           without -t default = 1
```
The description of the test itself is a bit unclear I think but this is how under it:
if the ```--thread``` flag is provided without any number then we should run the test on all threads that ```cyclictest``` allowed to run on (i.e. max_cpus) 
without ```--thread``` test should run only on 1 thread.
Now for the wrapper we do provide the ```--thread``` flag without any number, so we want all CPUs (i.e max_cpus).
In addition to that, I guess testing only a single thread on a cluster with quite some CPUs is quite not productive am I right?
now regarding if the flag should be a string or int, I can convert it to an int, if you think it's more clear and use other value than ```""```

Comment (fedepaol) [reply cnf-tests/pod-utils/cyclictest-runner/main.go:33]: The description is clear, but with go flags we don't have a way to express that, since I think you can't pass a flag without arguments (if you use `flags`, probably other libraries will allow that).

In the current implementation you always pass `-t` to the binary, so the code must be changed a bit to accomodate that or the comment of the threads flag must be adjusted.

File: cnf-tests/pod-utils/hwlatdetect-runner/main.go
Comment (Tal-or) [reply cnf-tests/pod-utils/hwlatdetect-runner/main.go:40]: From what I see, the tool only knows how to deal with only one measurement unit, meaning:
```5s``` is legal but ```5m5s``` and ```5``` isn't.
So not sure Duration is a good choice here because 5 minutes in duration will be translated to ```5m0s``` and the tool doesn't know how to deal with that, I'll try to translate everything into microsecond which is the most common measurement unit in this context

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (Tal-or) [reply cnf-tests/pod-utils/cyclictest-runner/main.go:33]: OK I think I understand now, so let's do the following:
In case no ```--thread``` flag is provided at all, we'll go with the default which is ```1 thread```
and in order to mimic the max_cpus the user shall provide ```0``` for all, is that works for you?

Comment (fedepaol) [reply cnf-tests/pod-utils/cyclictest-runner/main.go:33]: Sounds good.
-1 for max_cpus sounds also like a reasonable option

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (Tal-or) [reply cnf-tests/pod-utils/cyclictest-runner/main.go]: My mistake here, it must be a flag because duration can't support more than hours, and we need days as well

File: cnf-tests/pod-utils/cyclictest-runner/main.go
Comment (Tal-or) [reply cnf-tests/pod-utils/cyclictest-runner/main.go]: Done

**Comment by fromanirh** [comment]
/retest
this seems like the cluster just failed to go up

**Comment by fromanirh** [comment]
the failure in these two lanes is consistent. We need to dive into them.

**Comment by yuvalk** [comment]
> the failure in these two lanes is consistent. We need to dive into them.

shouldnt we move to 4.9 here
https://github.com/openshift-kni/cnf-features-deploy/blob/master/feature-configs/deploy/performance/operator_catalogsource.yaml#L11
?

**Comment by fromanirh** [comment]
> > the failure in these two lanes is consistent. We need to dive into them.
> 
> shouldnt we move to 4.9 here
> https://github.com/openshift-kni/cnf-features-deploy/blob/master/feature-configs/deploy/performance/operator_catalogsource.yaml#L11
> ?

Likely (silly me didn't check). But this means all, or at least most of them, the recent PRs should fail like this one is failing, right?

**Comment by fromanirh** [comment]
anyway: https://github.com/openshift-kni/cnf-features-deploy/pull/617

**Comment by fromanirh** [comment]
The failures are the same we see in https://github.com/openshift-kni/cnf-features-deploy/pull/617 - and they all seem unrelated to this PR.

**Comment by fromanirh** [comment]
/override ci/prow/e2e-gcp
For reasons stated in https://github.com/openshift-kni/cnf-features-deploy/pull/569#issuecomment-885518801 I'm overriding the failing lanes (because we can't just rebase on #617 
I volunteer to monitor the PR and fix any further failure should my assessment be proven wrong

**Comment by fromanirh** [comment]
> /override ci/prow/e2e-gcp
> For reasons stated in [#569 (comment)](https://github.com/openshift-kni/cnf-features-deploy/pull/569#issuecomment-885518801) I'm overriding the failing lanes (because we can't just rebase on #617
> I volunteer to monitor the PR and fix any further failure should my assessment be proven wrong

actually we need, because of the conflict we have now. And so: https://github.com/openshift-kni/cnf-features-deploy/pull/621

**Comment by Tal-or** [comment]
Not needed anymore. merged https://github.com/openshift-kni/cnf-features-deploy/pull/621 instead.

**Pull request closed without merging**
